Extract request builder

mikerhodes opened the pull request:
# What

The operations in Objective-Cloudant have a lot of copy and paste code
used to build requests. This commit introduces a builder for the
NSURLRequest used to make the request to Cloudant. The idea is to remove
boilerplate code from the operations, in order to bring each operation
closer to the ideal of having only the code it needs and no more.

# How

The builder takes a CDTCouchOperation in its constructor. The
CDTCouchOperation has been updated to expose information required to
build the HTTP request as properties. Subclasses are expected to
override these properties as required for their specific requests.
For example, all operations will override the `httpPath` property.
The builder reads this information into private properties at init
time which are later used to build the request.

# Testing

This is a refactor; no new tests were added. We could add some tests for the new class, but I'm not sure they're necessary.

# Reviewers

reviewer @rhyshort

rhyshort commented on ObjectiveCloudant/Operations/CDTCouchOperation.h:
```diff
@@ -128,6 +128,27 @@ typedef NS_ENUM(NSInteger, CDTObjectiveCloudantErrors) {
+
+ Should be overridden by subclasses to provide request body data.
+ */
+@property (nullable, nonatomic, strong) NSData *httpRequestBody;
```
These should all probably be read-only

rhyshort replied:
I also don't think we should expose these properties outside of the framework, since its currently internal implementation. If we make it a protocol as we discussed however, it would make nice extension points for people to add new operations if they need to use a Couch/Cloudant feature eg Search, Views etc.

mikerhodes commented:
@rhyshort I addressed your comments:

* Properties made readonly in 79cf597
* Moved all request properties to a protocol defining a delegate of CDTOperationRequestBuilder in bc46388 with cleanup in 747fb90.
* Fixed a place where I used properties rather than ivars in an init in f896038.

rhyshort commented on ObjectiveCloudant/HTTP/CDTOperationRequestBuilderDelegate.h:
```diff
@@ -0,0 +1,51 @@
+/**
+ Query items for this operation.
+ */
+- (NSArray<NSURLQueryItem *> *)queryItems;
```
needs nullability annotations

rhyshort commented:
+1

xpqz commented:
+1

mikerhodes commented:
Rewritten history, waiting for travis, then merging.
